File: go/errors_test.go
package tachyon

import (
	"context"
	"net/http"
	"testing"
	"time"
)

func TestErrorMappingByStatus(t *testing.T) {
	cases := []struct {
		status int
		code   string
		kind   ErrorKind
		check  func(error) bool
	}{
		{400, "invalid_query", ErrKindRequest, IsRequestError},
		{401, "unauthorized", ErrKindAuthentication, IsAuthenticationError},
		{403, "forbidden", ErrKindAuthorization, IsAuthorizationError},
		{404, "collection_not_found", ErrKindNotFound, IsNotFoundError},
		{409, "collection_exists", ErrKindConflict, IsConflictError},
		{500, "internal_error", ErrKindServer, IsServerError},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(tc.code, func(t *testing.T) {
			client, _ := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
				writeJSON(w, tc.status, `{"error":{"code":"`+tc.code+`","message":"boom: `+tc.code+`"}}`)
			})

			_, err := client.Collections.Retrieve(context.Background(), "products")
			if err == nil {
				t.Fatal("expected an error")
			}
			var tachyonErr *Error
			if !isTachyonError(err, &tachyonErr) {
				t.Fatalf("expected a *tachyon.Error, got %T: %v", err, err)
			}
			if tachyonErr.Code != tc.code || tachyonErr.Status != tc.status || tachyonErr.Kind != tc.kind {
				t.Fatalf("unexpected error: %+v", tachyonErr)
			}
			if !tc.check(err) {
				t.Fatalf("predicate for %s returned false", tc.code)
			}
		})
	}
}

func TestConnectionErrorOnNetworkFailure(t *testing.T) {
	// Nothing listens on port 1 (a reserved, unused TCP port), so this
	// exercises a real network failure rather than a mocked one.
	client := NewClient("http://127.0.0.1:1", WithTimeout(2*time.Second))

	_, err := client.Health(context.Background())
	if err == nil {
		t.Fatal("expected an error")
	}
	if !IsConnectionError(err) {
		t.Fatalf("expected a connection error, got %T: %v", err, err)
	}
	if IsTimeoutError(err) {
		t.Fatalf("did not expect a timeout error: %v", err)
	}
}

func TestTimeoutErrorWhenServerIsSlow(t *testing.T) {
	client, _ := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(200 * time.Millisecond)
		writeJSON(w, 200, `{"ok":true,"version":"x","uptime_seconds":0,"num_collections":0}`)
	})

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
	defer cancel()

	_, err := client.Health(ctx)
	if err == nil {
		t.Fatal("expected an error")
	}
	if !IsTimeoutError(err) {
		t.Fatalf("expected a timeout error, got %T: %v", err, err)
	}
	if !IsConnectionError(err) {
		t.Fatalf("expected IsConnectionError to also be true for a timeout: %v", err)
	}
}

func TestFallsBackToGenericErrorCodeWhenBodyIsNotTheDocumentedShape(t *testing.T) {
	client, _ := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(500)
		_, _ = w.Write([]byte("upstream exploded"))
	})

	_, err := client.Health(context.Background())
	var tachyonErr *Error
	if !isTachyonError(err, &tachyonErr) {
		t.Fatalf("expected a *tachyon.Error, got %T: %v", err, err)
	}
	if tachyonErr.Code != "internal_error" || tachyonErr.Status != 500 {
		t.Fatalf("unexpected error: %+v", tachyonErr)
	}
}

func isTachyonError(err error, target **Error) bool {
	e, ok := err.(*Error)
	if !ok {
		return false
	}
	*target = e
	return true
}

File: java/src/test/java/io/github/adikeshri/tachyon/ErrorsTest.java
package io.github.adikeshri.tachyon;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.io.IOException;
import java.time.Duration;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class ErrorsTest {

    static Stream<Arguments> statusToException() {
        return Stream.of(
            Arguments.of(400, "invalid_query", TachyonRequestException.class),
            Arguments.of(401, "unauthorized", TachyonAuthenticationException.class),
            Arguments.of(403, "forbidden", TachyonAuthorizationException.class),
            Arguments.of(404, "collection_not_found", TachyonNotFoundException.class),
            Arguments.of(409, "collection_exists", TachyonConflictException.class),
            Arguments.of(500, "internal_error", TachyonServerException.class));
    }

    @ParameterizedTest
    @MethodSource("statusToException")
    void mapsHttpStatusToExceptionType(int status, String code, Class<? extends TachyonException> expectedType) throws IOException {
        try (TestServer server = new TestServer(req -> TestServer.FakeResponse.json(status,
            "{\"error\":{\"code\":\"" + code + "\",\"message\":\"boom: " + code + "\"}}"))) {
            TachyonClient client = TestClients.forServer(server);

            TachyonException exception = assertThrows(expectedType, () -> client.collections.retrieve("products"));

            assertEquals(code, exception.getCode());
            assertEquals(status, exception.getStatusCode());
            assertEquals("boom: " + code, exception.getMessage());
        }
    }

    @Test
    void wrapsNetworkFailuresInConnectionException() {
        // Nothing listens on port 1 (a reserved, unused TCP port), so this
        // exercises a real network failure rather than a mocked one.
        TachyonClient client = TachyonClient.builder("http://127.0.0.1:1").timeout(Duration.ofSeconds(2)).build();

        assertThrows(TachyonConnectionException.class, client::health);
    }

    @Test
    void raisesTimeoutExceptionWhenServerIsSlow() throws IOException {
        try (TestServer server = new TestServer(req -> {
            try {
                Thread.sleep(200);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            return TestServer.FakeResponse.json(200, "{\"ok\":true,\"version\":\"x\",\"uptime_seconds\":0,\"num_collections\":0}");
        })) {
            TachyonClient client = TachyonClient.builder(server.baseUrl).timeout(Duration.ofMillis(10)).build();

            assertThrows(TachyonTimeoutException.class, client::health);
        }
    }

    @Test
    void fallsBackToGenericCodeWhenBodyIsNotTheDocumentedShape() throws IOException {
        try (TestServer server = new TestServer(req -> TestServer.FakeResponse.text(500, "upstream exploded"))) {
            TachyonClient client = TestClients.forServer(server);

            TachyonException exception = assertThrows(TachyonServerException.class, client::health);

            assertEquals("internal_error", exception.getCode());
            assertEquals(500, exception.getStatusCode());
        }
    }
}
